alembic: use threadsafe hdf5

guibou opened the pull request:
###### Motivation for this change

Allowing `alembic` (and perhaps other libraries) to load `hdf5` encoded files in a concurrent manner. To achieve this, this change:

- Enables thread-safe support in `hdf5` in a separate package `hdf5-threadsafe`.
- Use the later for the `alembic` package

###### Discussion

By default, `hdf5` is not thread-safe hence `--enable-threadsafe` is mandatory to load `alembic` file (using `hdf5`) in a concurrent context.

However this change is not benign for all `hdf5` users: the High Level (`hl`) library included in `hdf5` does not officially support thread safety. The recommended way is to `--disable-hl`. However this may not be acceptable for some users of hdf5 (actually 21 derivation in nixpkgs). So I do not propose this for the default hdf5 derivation.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

FRidh commented:
- If I am correct `zlib` support is currently enabled.
- `szip` is listed as unfree. Does alembic need `szip` support to function? If not, then it should not use `szip` by default.

guibou commented:
@FRidh 

- Yes, you are right, `zlib` is currently enabled. I assumed that `callPackage` does not fill the value with a default, such as `zlib = null`, but I'm apparently wrong.
- Alembic works without `szip`, you are right, and I understand the issue about the `szip` license, so I won't complain and override locally my project to get `szip`.

So the issue about `szip` and `zlib` can be closed. What about hdf5 and threading?

peterhoeg commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13649,7 +13656,9 @@ with pkgs;
-  alembic = callPackage ../development/libraries/alembic {};
+  alembic = callPackage ../development/libraries/alembic {
+    hdf5 = hdf5-threadsafe;
+  };
```
Since ```hdf5-threadsafe``` is in all-packages.nix why not just pass that in directly instead of overriding hdf5? This makes things more explicit for alembic and easier to understand.

guibou replied:
You mean to directly use `hdf5-threadsafe` inside `alembic`? Yes, that's a solution. I'll update the PR with this.

guibou commented:
Any update? I'm using this (2 lines) changes in production since October without any issue, can we merge?

peterhoeg commented on pkgs/development/libraries/alembic/default.nix:
```diff
@@ -14,7 +14,7 @@ stdenv.mkDerivation rec
   outputs = [ "bin" "dev" "out" "lib" ];
 
-  buildInputs = [ unzip cmake openexr hdf5 ];
+  buildInputs = [ unzip cmake openexr hdf5-threadsafe ];
```
Please shift unzip and cmake to ```nativeBuildInputs```.

guibou replied:
Done.

guibou committed:
hdf5: introduce hdf5-threadsafe variant

Compiled with thread safe support and without the High Level library
which is incompatible with thread safety.

guibou committed:
alembic: clean buildInputs to nativeBuildInputs

guibou committed:
alembic: use hdf5-threadsafe

peterhoeg commented:
Thanks @guibou !
